### PR TITLE
Replaced Whoosh with ElasticSearch

### PR DESCRIPTION
**Pull request by shreyas-satish**
Using [pyelasticsearch](https://pyelasticsearch.readthedocs.org/en/latest/) to connect to Elastic Search.

For Reference:
http://www.elasticsearch.org/guide/en/elasticsearch/guide/current/scoring-theory.html
http://www.elasticsearch.org/guide/en/elasticsearch/reference/current/analysis-synonym-tokenfilter.html

### CODE REVIEW

File: hasjob/models/jobpost.py [hunk 1]
```diff
@@ -181,6 +181,10 @@ class JobPost(BaseMixin, db.Model):
     def get(cls, hashid):
         return cls.query.filter_by(hashid=hashid).one_or_none()
 
+    @classmethod
+    def get_by_ids(cls, ids):
+        return cls.query.filter(cls.id.in_(ids)).all()
```
Comment (jace) [inline hasjob/models/jobpost.py]: You don't want the entire content of the jobpost here. Use `.options(db.load_only(column_name, ...))` to limit what you load.

File: hasjob/search.py [hunk 2]
```diff
@@ -1,83 +1,45 @@
...
-from sqlalchemy.exc import ProgrammingError
-from whoosh import fields, index
-from whoosh.index import LockError
-from whoosh.qparser import QueryParser
-from whoosh.analysis import StemmingAnalyzer
-
 from hasjob import models, app
 
+from pyelasticsearch import ElasticSearch
+es = ElasticSearch('http://localhost:9200/')
```
Comment (jace) [inline hasjob/search.py]: Move this to `hasjob/__init__.py` and don't hardcode the URL here. Load from config in `init_for`.

File: hasjob/search.py [hunk 3]
```diff
@@ -1,83 +1,45 @@
...
-
-    if expand:
-        return (ob_from_idref(hit['idref']) for hit in hits)
-    else:
-        return hits
+        hits = es.search(query, index=ES_INDEX)
+        ids = map(lambda hit: id_from_idref(hit[u'_id']), hits['hits']['hits'])
+        job_post_ids = map(lambda post: post.id, models.JobPost.get_by_ids(ids))
+        # Handles the event of the index returning an id that isn't present in the DB
+        return [models.JobPost.query.get(id) for id in filter(lambda id: id in job_post_ids, ids)]
```
Comment (jace) [inline hasjob/search.py]: I can't make out what is going on here.

File: hasjob/search.py [hunk 4]
```diff
@@ -1,83 +1,45 @@
...
-from whoosh.index import LockError
-from whoosh.qparser import QueryParser
-from whoosh.analysis import StemmingAnalyzer
-
 from hasjob import models, app
 
+from pyelasticsearch import ElasticSearch
+es = ElasticSearch('http://localhost:9200/')
+
+INDEXABLE = (models.JobPost,)
```
Comment (jace) [inline hasjob/search.py]: Are we not indexing `JobType`, `JobCategory` and `Tag`? We'll have more document types soon: `Domain`, `JobApplication`, `Organization` and more.

File: hasjob/search.py [hunk 5]
```diff
@@ -95,31 +57,16 @@ def on_models_committed(sender, changes):
...
-                for ob in model.query.all():
-                    mapping = ob.search_mapping()
-                    public = mapping.pop('public')
-                    if public:
-                        writer.add_document(**mapping)
-            except ProgrammingError:
-                pass  # The table doesn't exist yet. This is really a new installation.
-        writer.commit()
+    for model in INDEXABLE:
+        records = map(lambda record: record.search_mapping(), model.query.all())
```
Comment (jace) [inline hasjob/search.py]: This will choke on production. It doesn't have enough RAM to do `JobPost.query.all()`.

**Comment by shreyas-satish** [comment]
```python
ids = map(lambda hit: id_from_idref(hit[u'_id']), hits['hits']['hits'])
```
Maps over the 'hits' list, extracts the id from the idref from each hit (example idref: 'post/20') and returns a list of these ids.

```python
job_post_ids = map(lambda post: post.id, models.JobPost.get_by_ids(ids))
```
Maps over the list of job posts, and returns a list of only the ids

```python
return [models.JobPost.query.get(id) for id in filter(lambda id: id in job_post_ids, ids)]
```
Filters the original ids list down to a list of ids that were found in the database, and then returns job posts for these ids.

If readability is a problem here, I'll change the do_search method to use loops instead.

**Comment by jace** [comment]
Use list comprehensions. That'll fix it.

File: hasjob/search.py [hunk 6]
```diff
@@ -1,83 +1,45 @@
...
-            searcher = ix.searcher()
-            hits = searcher.search(qry, limit=1000)
-
-    if expand:
-        return (ob_from_idref(hit['idref']) for hit in hits)
-    else:
-        return hits
+        hits = es.search(query, index=ES_INDEX)
+        ids = map(lambda hit: id_from_idref(hit[u'_id']), hits['hits']['hits'])
+        job_post_ids = map(lambda post: post.id, models.JobPost.get_by_ids(ids))
```
Comment (jace) [inline hasjob/search.py]: Isn't this exactly the same set of ids as in the previous line? Why hit the database at all?

Comment (jace) [inline hasjob/search.py]: To make this efficient:

```python
job_post_ids = [r.id for r in db.session.query(models.JobPost.id).filter(models.JobPost.id.in_(ids))]
```

See? Doesn't involve a map+lambda (inefficient) or loading unwanted data from the database.

**Comment by jace** [comment]
For some reason this PR is full of commits that are already in master.

**Comment by shreyas-satish** [comment]
I updated the 'es' branch with the changes in master last night.

**Comment by jace** [comment]
Fixed after I merged with master again.

**Comment by shreyas-satish** [comment]
Closing PR. This branch will stay, just in case it's required in the future.

**Pull request closed without merging**